langchain4j-agentic/src/main/java/dev/langchain4j/agentic/cognisphere/Cognisphere.java

```diff
@@ -0,0 +1,233 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Predicate;
+
+/**
+ * The Cognisphere class represents a cognitive environment where agents belonging to the same
+ * agentic system can share their state.
+ * It maintains the state of the environment, tracks agent invocations, and provides
+ * methods to allow agents to interact with the shared state.
+ * <p>
+ * Agents can register their calls, and the context of interactions is stored for later retrieval.
+ * The class also provides methods to read and write state, manage agent invocations, and retrieve
+ * the context as a conversation.
+ */
+public class Cognisphere {
+
+    private static final Logger LOG = LoggerFactory.getLogger(Cognisphere.class);
+
+    public record AgentMessage(String agentName, ChatMessage message) {}
+
+    private final CognisphereKey key;
+    private final Map<String, Object> state = new ConcurrentHashMap<>();
+    private final Map<String, List<AgentCall>> agentsInvocations = new ConcurrentHashMap<>();
+    private final List<AgentMessage> context = Collections.synchronizedList(new ArrayList<>());
+
+    public enum Kind {
+        EPHEMERAL, REGISTERED, PERSISTENT
+    }
+    private final Kind kind;
+    private transient ReadWriteLock lock = null;
+
+    Cognisphere(Kind kind) {
+        this(new CognisphereKey("", UUID.randomUUID().toString()), kind);
+    }
+
+    Cognisphere(CognisphereKey key, Kind kind) {
+        this.key = key;
+        this.kind = kind;
+        if (kind == Kind.PERSISTENT) {
+            lock = new ReentrantReadWriteLock();
+        }
+    }
+
+    public CognisphereKey key() {
+        return key;
+    }
+
+    public Object id() {
+        return key.id();
+    }
+
+    public void writeState(String key, Object value) {
+        withReadLock(() -> {
+            if (value == null) {
+                state.remove(key);
+            } else {
+                state.put(key, value);
+            }
+        });
+    }
+
+    public void writeStates(Map<String, Object> newState) {
+        withReadLock(() -> state.putAll(newState));
+    }
+
+    public boolean hasState(String key) {
+        Object value = state.get(key);
+        if (value == null) {
+            return false;
+        }
+        return value instanceof String s ? !s.isBlank() : true;
+    }
+
+    public Object readState(String key) {
+        return state.get(key);
+    }
+
+    public <T> T readState(String key, T defaultValue) {
+        return (T) state.getOrDefault(key, defaultValue);
+    }
+
+    public Map<String, Object> getState() {
+        return state;
+    }
+
+    public void registerAgentCall(AgentSpecification agentSpecification, Object agent, Object[] input, Object response) {
```

Comment:
```suggestion
    public void registerAgentCall(AgentSpecification agentSpecification, Object agent, Object[] input, Object output) {
```
nit: for consistency in names: input/output

Reply:
Fixed.